From 8c4445d4dbd4c2d5de5a19370fcc5d8777a7e6d4 Mon Sep 17 00:00:00 2001
From: Andrew Davis <afd@ti.com>
Date: Fri, 6 Mar 2020 13:06:55 -0500
Subject: [PATCH] backend-drm: Select plane based on current attached CRTC

When doing plane selection for an output CRTC check if the plane
already has a CRTC attached and if so prefer that plane only for
the corresponding CRTC.

This prevents changing a CRTC's primary plane when it is active
which is not allowed by the DRM framework.

Upstream-Status: Backport 94afcbcdc3f1b1cfc050da242c9c70009b007fc6

Based-on-patch-by: Eric Ruei <e-ruei1@ti.com>
Signed-off-by: Andrew Davis <afd@ti.com>
---
 libweston/backend-drm/drm-internal.h | 1 +
 libweston/backend-drm/drm.c          | 9 +++++++++
 2 files changed, 10 insertions(+)

diff --git a/libweston/backend-drm/drm-internal.h b/libweston/backend-drm/drm-internal.h
index 48600880..b381ca31 100644
--- a/libweston/backend-drm/drm-internal.h
+++ b/libweston/backend-drm/drm-internal.h
@@ -468,6 +468,7 @@ struct drm_plane {
 	uint32_t possible_crtcs;
 	uint32_t plane_id;
	uint32_t plane_idx;
+	uint32_t crtc_id;
 
 	struct drm_property_info props[WDRM_PLANE__COUNT];

diff --git a/libweston/backend-drm/drm.c b/libweston/backend-drm/drm.c
index 42787702..b5b7c4ba 100644
--- a/libweston/backend-drm/drm.c
+++ b/libweston/backend-drm/drm.c
@@ -776,6 +776,7 @@ drm_plane_create(struct drm_backend *b, const drmModePlane *kplane)
	plane->state_cur->complete = true;
	plane->possible_crtcs = kplane->possible_crtcs;
	plane->plane_id = kplane->plane_id;
+	plane->crtc_id = kplane->crtc_id;

	weston_drm_format_array_init(&plane->formats);
 
@@ -866,6 +867,14 @@ drm_output_find_special_plane(struct drm_backend *b, struct drm_output *output,
 		if (found_elsewhere)
 			continue;
 
+		/* If a plane already has a CRTC selected and it is not our
+		 * output's CRTC, then do not select this plane. We cannot
+		 * switch away a plane from a CTRC when active. */
+		if ((type == WDRM_PLANE_TYPE_PRIMARY) &&
+		    (plane->crtc_id != 0) &&
+		    (plane->crtc_id != output->crtc->crtc_id))
+			continue;
+
		plane->possible_crtcs = (1 << output->crtc->pipe);
 		return plane;
 	}
-- 
2.38.1

